validation_rules.js

```diff
@@ -0,0 +1,1258 @@
+    _get_wall_area_left_right() {
+        let length = this._get_wall_dimension_left_right();
+        let height = parseInt(this._homeValues.floor_to_ceiling_height) || false;
+        let stories = parseInt(this._homeValues.num_floor_above_grade) || false;
+        if (length && height && stories) {
+            return parseInt(length * height * stories);
+        } else {
+            return false;
+        }
+    }
+
+    /*
+     * Checks that the combined roof_area is not less than the combined floor_area
+     */
+    _check_combined_area() {
+        let combinedRoofCeilingArea = this._get_combined_roof_ceiling_area();
+        let combinedFloorArea = this._get_combined_floor_area();
+        if (combinedRoofCeilingArea  <= combinedFloorArea * .95) { // Allow 5% error
+            return "The roof does not cover the floor";
+        } else {
+            return false;
+        }
+    }
+
+    /*
+     * Checks that the roof_area and floor_areas are consistent with conditioned footprint areas
+     * @param {number} combinedArea
+     * @param {'roof'|'floor'} thisAreaType
+     */
+    _check_conditioned_areas(combinedArea, thisAreaType) {
+        let footprintArea = this._get_footprint_area();
+        if (TypeRules._int_or_zero(this._homeValues.num_floor_above_grade) === 0) {
+            return "This home’s minimum footprint is unknown.  Please enter number of stories.";
+        } else {
+            // Check that combined areas are within reasonable range of footprint
+            // const max = thisAreaType === "roof"
+            //     ? footprintArea * 2.5 // roof area max
+            //     : this._homeValues.conditioned_floor_area * 1.05; // floor area & ceiling area max
+            const max = footprintArea * 2.5;
+            const expectedRange = [footprintArea * 0.95, max];
+            if (!((expectedRange[0] < combinedArea) && (combinedArea < expectedRange[1]))) {
+                return `
+                    This home's minimum footprint is approximately ${footprintArea}sqft, but you
+                    have specified ${combinedArea}sqft of total ${thisAreaType} area. The allowed range
+                    is (${Math.ceil(expectedRange[0])}sqft - ${Math.floor(expectedRange[1])}sqft).
+                    Please adjust any incorrect values. *The footprint is calculated as
+                    (<total area> - <conditioned basement area>) / <number of floors>
+                `;
+            }
+        }
+    }
+
+    /*
+     * Checks address/city or coordinates to be an accepted input.
+     * Intended to validate half addresses that do not require address and/or city feilds, if coordinates are provided on map.
+     * @param {string} value
+     */
+    _require_if_no_coordinates(value) {
+        if((this._homeValues.latitude && this._homeValues.longitude) || value){
+            return null;
+        }
+        else{
+            return "Field is required, if no coordinates are provided."
+        }
+    }
+};
```

Comment:
This is basically the previous tool without changes, correct?

Reply:
Correct, validation_rules.js had minimal changes so as to not impact the "flat format" validation.